msprime/intervals.py

```diff
@@ -449,50 +449,151 @@ def asdict(self):
         return self.map.asdict()
 
 
-def read_hapmap(filename):
+def read_hapmap(
+    fileobj,
+    sequence_length=None,
+    *,
+    has_header=True,
+    position_col=None,
+    rate_col=None,
+    map_col=None,
+):
     # Black barfs with an INTERNAL_ERROR trying to reformat this docstring,
     # so we explicitly disable reformatting here.
     # fmt: off
     """
     Parses the specified file in HapMap format. These files must be
-    white-space-delimited, and contain a single header line (which is
-    ignored), and then each subsequent line contains the starting position
-    and recombination rate for the segment from that position (inclusive)
-    to the starting position on the next line (exclusive). Starting
-    positions of each segment are given in units of bases, and
-    recombination rates in centimorgans/Megabase. The first column in this
-    file is ignored, as are additional columns after the third (Position is
-    assumed to be the second column, and Rate is assumed to be the third).
-    If the first starting position is not equal to zero, then a
-    zero-recombination region is inserted at the start of the chromosome.
-
-    A sample of this format is as follows::
-
-        Chromosome	Position(bp)	Rate(cM/Mb)	Map(cM)
-        chr1	55550	        2.981822	0.000000
-        chr1	82571	        2.082414	0.080572
-        chr1	88169	        2.081358	0.092229
-        chr1	254996	        3.354927	0.439456
-        chr1	564598	        2.887498	1.478148
+    white-space-delimited, and by default are assumed to contain a single header line
+    (which is ignored). Each subsequent line then contains a physical position (in base
+    pairs) and either a genetic map position (in centiMorgans) or a recombination rate
+    (in centiMorgans per megabase); the rate between the current physical position
+    (inclusive) and the physical position on the next line (exclusive) is taken as
+    constant. By default, the second column of the file is taken as the physical
+    position and the fourth column is taken as the genetic position, as seen in the
+    following sample of the format::
+
+        Chromosome	Position(bp)  Rate(cM/Mb)  Map(cM)
+        chr10       48232         0.1614       0.002664
+        chr10       48486         0.1589       0.002705
+        chr10       50009         0.159        0.002947
+        chr10       52147         0.1574       0.003287
         ...
-        chr1	182973428	2.512769	122.832331
-        chr1	183630013	0.000000	124.482178
-
-    :param str filename: The name of the file to be parsed. This may be
-        in plain text or gzipped plain text.
+        chr10	    133762002     3.358        181.129345
+        chr10	    133766368     0.000        181.144008
+
+    :: note..
```

Comment:
What if the column at position zero contained something completely different, wouldn't this warning be unnecessary and annoying then? There's no need for this check, it's not doing anything useful *unless* we actually use the value at some point.

Reply:
That's a good point. The reason I want to keep the check is that (a) it is common to munge multiple chromosomes together in the same file, and we should check that the user is not doing this by mistake (b) it is useful to have the chromosome name, and we had thought that we might use it in future (although we are not doing so at the moment) and (c) @grahamgower suggests (and I agree) that we should try to be pretty strict about the format we are defaulting to in read_hapmap(), so that we don't get into the bind we had before of making files that will "work" in stdpopsim, but are not in any conventional format.

My inclination therefore is to keep it, but maybe to have a keyword only param that either specifies the column containing the name (default=0) or if a string, the chromosome name. I suspect that you won't like this though?

Reply:
Let's drop if for now, please. It'll break anyway if someone puts in multiple chrs because of the ordering requirement of positions.

Reply:
OK. I'll open up an issue and paste the code & test in there so we don't forget it, though.

Reply:
Now in https://github.com/tskit-dev/msprime/issues/1363